common-pg/src/main/scala/com/socrata/Predicates.scala

```diff
@@ -0,0 +1,29 @@
+package com.socrata
+
+object Predicates {
+
+  type Predicate[T] = Function[T, Boolean]
+
+  implicit class PredicateOperators[T](me: Predicate[T]) {
+
+    def &(them: Predicate[T]): Predicate[T] = (t: T) => me(t) & them(t)
+
+    def &&(them: Predicate[T]): Predicate[T] = (t: T) => me(t) && them(t)
+
+    def |(them: Predicate[T]): Predicate[T] = (t: T) => me(t) | them(t)
+
+    def ||(them: Predicate[T]): Predicate[T] = (t: T) => me(t) || them(t)
+
+    // this allows you to specify negation of a predicate before its applied
+    // e.g (!isAlphaNumeric && !isUnderscore)('a')
+    def unary_!(c: T): Boolean = !me(c)
```

Comment:
This fn cannot be invoked.
You're demanding a T in order to invoke this ! fn. That means you'll need to provide an argument to invoke !.
You won't be able to write !foo && !bar.

I'd recommend using this pattern https://coderwall.com/p/k_1jzw/scala-s-pimp-my-library-pattern-example
which in your case would be
```
package com.socrata

import scala.language.implicitConversions

object Predicates {

  case class Predicate[T](fn: Function[T, Boolean]) { self =>
    def apply(t: T) = fn(t)
    def ||(them: Predicate[T]): Predicate[T] = Predicate(t => self(t) || them(t))
    def &&(them: Predicate[T]): Predicate[T] = Predicate(t => self(t) && them(t))
    def unary_!(): Predicate[T] = Predicate(t => !self(t))
  }

  implicit def fnTopredicate[T](fn: T => Boolean): Predicate[T] = Predicate(fn)
  implicit def predicateToFn[T](pred: Predicate[T]) = pred.fn

  def isAlphaNumericUnderscore: Predicate[Char] = isAlphaNumeric || isUnderscore

  def isAlphaNumeric: Predicate[Char] = Character.isLetterOrDigit _

  def isUnderscore: Predicate[Char] = (_: Char) == '_'

}
```

Reply:
Oh you are right that `!something && !other` fails.
The solution you pasted works. 
Ill update this.

Reply:
Is it possible to do this without a wrapping object on the function?

Reply:
it'll only work for predicates, not all functions, and only where this is imported.

Reply:
I think the PML pattern can get pretty scary pretty quickly. There's lots of hidden conversions that can occur and chew through your performance.
I'm not sure of another way around this.

Reply:
btw scala3 has a much nicer mechanism for this.

Reply:
Okay ill need a minute on this one, I'm not happy with wrapping the function in an object

Reply:
Interesting...it can work, BUT
It comes down to `unary_!` needs to be a method(val) instead of a function(def).
Further complicated by scala casting between function and methods, and generics.

Reply:
If we leave generics in place, but turn unary_! into a method, then ->
```
value || is not a member of Char => Boolean
(!isAlphaNumeric || isUnderscore)('!') 
```

Reply:
If you turn `unary_!` into a method, but replace generic Predicate[T] and (Char)=>Boolean, it works out.

Reply:
Ah the key is that the `me` of the implicit needs to be completely unwrapped and raw
```
implicit class PredicateOperators[T](me: (T)=>Boolean) 
```

Reply:
```
type Predicate[T] = (T)=>Boolean

implicit class PredicateOperators[T](me: (T)=>Boolean) {

  def &(them: Predicate[T] ): Predicate[T]  = (t: T) => me(t) & them(t)

  def &&(them: Predicate[T] ): Predicate[T]  = (t: T) => me(t) && them(t)

  def |(them: Predicate[T] ): Predicate[T]  = (t: T) => me(t) | them(t)

  def ||(them: Predicate[T] ): Predicate[T]  = (t: T) => me(t) || them(t)

  // this allows you to specify negation of a predicate before its applied
  // e.g (!isAlphaNumeric && !isUnderscore)('a')
  val unary_! = (c:T) => !me(c)

}

def isAlphaNumeric: Predicate[Char] = Character.isLetterOrDigit

def isUnderscore: Predicate[Char] = _ == '_'

def isAlphaNumericUnderscore: Predicate[Char] = isAlphaNumeric || isUnderscore

(isAlphaNumeric || isUnderscore)('a')
(isAlphaNumeric || isUnderscore)('_')
(isAlphaNumeric || isUnderscore)('!')
(isAlphaNumeric || !isUnderscore)('!')
(!isAlphaNumeric || isUnderscore)('!')
(!isAlphaNumeric || !isUnderscore)('a')

```